platform/linux-generic/odp_ipsec.c

```diff
@@ -1963,6 +1963,81 @@ int odp_ipsec_out_inline(const odp_packet_t pkt_in[], int num_in,
 	return in_pkt;
 }
 
+static void ipsec_out_sa_info(ipsec_sa_t *ipsec_sa, odp_ipsec_sa_info_t *sa_info)
+{
+	sa_info->outbound.seq_num =
+		(uint64_t)odp_atomic_load_u64(&ipsec_sa->hot.out.seq) -	1;
+
+	if (ipsec_sa->param.mode == ODP_IPSEC_MODE_TUNNEL) {
+		uint8_t *src, *dst;
+
+		if (ipsec_sa->param.outbound.tunnel.type ==
+				ODP_IPSEC_TUNNEL_IPV4) {
+			src = sa_info->outbound.tunnel.ipv4.src_addr;
+			dst = sa_info->outbound.tunnel.ipv4.dst_addr;
+			memcpy(src, &ipsec_sa->out.tun_ipv4.src_ip,
+			       _ODP_IPV4ADDR_LEN);
+			memcpy(dst, &ipsec_sa->out.tun_ipv4.dst_ip,
+			       _ODP_IPV4ADDR_LEN);
+			sa_info->param.outbound.tunnel.ipv4.src_addr = src;
+			sa_info->param.outbound.tunnel.ipv4.dst_addr = dst;
+		} else {
+			src = sa_info->outbound.tunnel.ipv6.src_addr;
+			dst = sa_info->outbound.tunnel.ipv6.dst_addr;
+			memcpy(src, &ipsec_sa->out.tun_ipv6.src_ip,
+			       _ODP_IPV6ADDR_LEN);
+			memcpy(dst, &ipsec_sa->out.tun_ipv6.dst_ip,
+			       _ODP_IPV6ADDR_LEN);
+			sa_info->param.outbound.tunnel.ipv6.src_addr = src;
+			sa_info->param.outbound.tunnel.ipv6.dst_addr = dst;
+		}
+	}
+}
+
+static void ipsec_in_sa_info(ipsec_sa_t *ipsec_sa, odp_ipsec_sa_info_t *sa_info)
+{
+	if (ipsec_sa->param.mode == ODP_IPSEC_MODE_TUNNEL) {
+		uint8_t *dst = sa_info->inbound.lookup_param.dst_addr;
+
+		if (ipsec_sa->param.inbound.lookup_param.ip_version ==
+		    ODP_IPSEC_IPV4)
+			memcpy(dst, &ipsec_sa->in.lookup_dst_ipv4,
+			       _ODP_IPV4ADDR_LEN);
+		else
+			memcpy(dst, &ipsec_sa->in.lookup_dst_ipv6,
+			       _ODP_IPV6ADDR_LEN);
+
+		sa_info->param.inbound.lookup_param.dst_addr = dst;
+	}
+
+	if (ipsec_sa->antireplay) {
+		sa_info->inbound.antireplay_ws = IPSEC_ANTIREPLAY_WS;
+		sa_info->inbound.antireplay_window_top =
+			_odp_ipsec_sa_antireplay_max_seq(ipsec_sa);
+	}
+}
+
+int odp_ipsec_sa_info(odp_ipsec_sa_t sa, odp_ipsec_sa_info_t  *sa_info)
```

Comment:
Maybe a better place for this API function and its callees would be odp_ipsec_sad.c where SA creation, anti-replay window update etc reside?

Reply:
Agreed. Moved the functions to odp_ipsec_sad.c.